pysph/solver/tools.py

```diff
@@ -81,9 +81,78 @@ def __init__(self, app, array_name, props, freq=100, xi=None, yi=None,
         )
 
     def post_step(self, solver):
-        if solver.count%self.freq == 0 and solver.count > 0:
+        if solver.count % self.freq == 0 and solver.count > 0:
             self.interp.nnps.update()
             data = dict(x=self.xi, y=self.yi, z=self.zi)
             for prop in self.props:
                 data[prop] = self.interp.interpolate(prop)
             self.array.set(**data)
+
+
+class DensityCorrection(Tool):
+    """
+    A tool to reinitialize the density of the fluid particles
+    """
+
+    def __init__(self, app, arr_names, corr='shepard', freq=10, kernel=None):
+        """
+        Parameters
+        ----------
+
+        app : pysph.solver.application.Application.
+            The application instance.
+        arr_names : array
+            Names of the particle arrays whose densities needs to be
+            reinitialized.
+        corr : str
+            Name of the density reinitialization operation
+        freq : int
+            Frequency of reinitialization.
+        kernel: any kernel from pysph.base.kernels
+
+        """
+        self.freq = freq
+        self.corr = corr
+        self.names = arr_names
+        self.count = 1
+        self._sph_eval = None
+        self.kernel = kernel
+        self.dim = app.solver.dim
+        self.particles = app.particles
+
+    def _get_sph_eval_shepard(self):
+        from pysph.sph.wc.density_correction import (ShepardFilterPreStep,
+                                                     ShepardFilter)
+        from pysph.tools.sph_evaluator import SPHEvaluator
+        from pysph.solver.utils import get_array_by_name
+        from pysph.sph.equation import Group
+        if self._sph_eval is None:
+            arrs = [get_array_by_name(self.particles, i) for i in self.names]
+            eqns = []
+            for arr in arrs:
+                arr.add_property('tw')
```

Comment:
Add the property only if it is not used already.  Also maybe use something else instead of `tw`, maybe `tmp_wt`?